File: web/cache/src/main/java/me/spring/web/cache/order/OrderService.java
package me.spring.web.cache.order;

import org.springframework.cache.annotation.Cacheable;
import org.springframework.stereotype.Service;

@Service
class OrderService {
    private final OrderRepository repository;

    OrderService(OrderRepository repository) {
        this.repository = repository;
    }

    @Cacheable(value = "orderNumber")
    public long getOrderAmount(long orderNumber) {
        return repository.getOrderAmount(orderNumber);
    }

}

File: web/oop-and-pp/src/main/kotlin/me/spring/web/oopandpp/domainmodel/application/OrderService.kt
package me.spring.web.oopandpp.domainmodel.application

import me.spring.web.oopandpp.domainmodel.domain.Order
import me.spring.web.oopandpp.domainmodel.infrastructure.DiscountStrategyRepository
import me.spring.web.oopandpp.domainmodel.infrastructure.OrderRepository
import me.spring.web.oopandpp.domainmodel.infrastructure.ProductRepository
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class OrderService(
    private val discountStrategyRepository: DiscountStrategyRepository,
    private val orderRepository: OrderRepository,
    private val productRepository: ProductRepository,
) {

    @Transactional
    fun order(order: Order) {
        val product = productRepository.findByIdOrNull(order.productId) ?: throw IllegalArgumentException("no exists product productId: ${order.productId}")
        val discountStrategies = discountStrategyRepository.findAllByProductId(product.id)
        order.applyDiscountFee(discountStrategies)
        orderRepository.save(order)
    }

}
